executor/slow_query.go

```diff
@@ -97,50 +101,48 @@ func dataForSlowLog(ctx sessionctx.Context) ([][]types.Datum, error) {
 		}
 		return true
 	}
-	return parseSlowLogFile(ctx.GetSessionVars().Location(), ctx.GetSessionVars().SlowQueryFile, checkValid)
-}
-
-// parseSlowLogFile uses to parse slow log file.
-// TODO: Support parse multiple log-files.
-func parseSlowLogFile(tz *time.Location, filePath string, checkValid checkValidFunc) ([][]types.Datum, error) {
-	file, err := os.Open(filePath)
+	rows, fileLine, err := ParseSlowLog(ctx, bufio.NewReader(e.file), e.fileLine, 1024, checkValid)
 	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	defer func() {
-		if err = file.Close(); err != nil {
-			logutil.BgLogger().Error("close slow log file failed.", zap.String("file", filePath), zap.Error(err))
+		if err == io.EOF {
+			e.retrieved = true
+		} else {
+			return nil, err
 		}
-	}()
-
-	return ParseSlowLog(tz, bufio.NewReader(file), checkValid)
+	}
+	e.fileLine = fileLine
+	if e.table.Name.L == strings.ToLower(infoschema.ClusterTableSlowLog) {
+		return infoschema.AppendHostInfoToRows(rows)
+	}
+	return rows, nil
 }
 
 type checkValidFunc func(string) bool
 
 // ParseSlowLog exports for testing.
 // TODO: optimize for parse huge log-file.
-func ParseSlowLog(tz *time.Location, reader *bufio.Reader, checkValid checkValidFunc) ([][]types.Datum, error) {
+func ParseSlowLog(ctx sessionctx.Context, reader *bufio.Reader, fileLine, maxRow int, checkValid checkValidFunc) ([][]types.Datum, int, error) {
 	var rows [][]types.Datum
 	startFlag := false
+	lineNum := fileLine
+	tz := ctx.GetSessionVars().Location()
 	var st *slowQueryTuple
-	lineNum := 0
 	for {
+		if len(rows) >= maxRow {
```

Comment:
when a file is in rotate, the last log in the tail of the file may be not complete. 
Part of it may be at the beginning of file.
when it `startFlag` is true, but got `EOF` at the tail of the file, should we continue to read backwards to finish it?